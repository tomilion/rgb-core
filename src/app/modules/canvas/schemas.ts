export enum CanvasState {
    PENDING = 0,
    ACTIVE = 1,
    COMPLETE = 2,
}

export interface CanvasPayload {
    ownerId: Buffer;
    costPerPixel: bigint;
    startTime: bigint;
    endTime: bigint;
    width: number;
    height: number;
    timeBetweenDraws: number;
    seed: bigint;
    state: number;
}

export const canvasSchema = {
    $id: "canvas/canvas",
    type: "object",
    required: ["ownerId", "costPerPixel", "startTime", "endTime", "width", "height", "timeBetweenDraws", "seed"],
    properties: {
        ownerId: { fieldNumber: 1, dataType: "bytes", minLength: 20, maxLength: 20 },
        costPerPixel: { fieldNumber: 2, dataType: "uint64" },
        startTime: { fieldNumber: 3, dataType: "uint64" },
        endTime: { fieldNumber: 4, dataType: "uint64" },
        width: { fieldNumber: 5, dataType: "uint32" },
        height: { fieldNumber: 6, dataType: "uint32" },
        timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
        seed: { fieldNumber: 8, dataType: "uint64" },
        state: { fieldNumber: 9, dataType: "uint32", default: CanvasState.PENDING },
    },
};

export interface PixelPayload {
    ownerId: Buffer;
    colour: number;
}

export const pixelSchema = {
    $id: "canvas/pixel",
    type: "object",
    required: ["ownerId", "colour"],
    properties: {
        ownerId: { fieldNumber: 1, dataType: "bytes", minLength: 20, maxLength: 20 },
        colour: { fieldNumber: 2, dataType: "uint32" },
    },
};

export interface AccountPayload {
    lastBlockHeight: number;
}

export const accountSchema = {
    $id: "canvas/account",
    type: "object",
    required: ["lastBlockHeight"],
    properties: {
        lastBlockHeight: { fieldNumber: 1, dataType: "uint32" },
    },
};

export interface PendingPayload {
    canvasIds: number[];
}

export const pendingSchema = {
    $id: "canvas/pending",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface ActivePayload {
    canvasIds: number[];
}

export const activeSchema = {
    $id: "canvas/active",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface CompletePayload {
    canvasIds: number[];
}

export const completeSchema = {
    $id: "canvas/complete",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};
