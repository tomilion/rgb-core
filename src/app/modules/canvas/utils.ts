export const serialiseCanvasId = (canvasId: number) => `canvas-${canvasId}`;

export const serialisePixelId = (canvasId: number, xCoord: number, yCoord: number) =>
    `${serialiseCanvasId(canvasId)}-pixel-${xCoord}-${yCoord}`;

export const serialiseAccountId = (canvasId: number, address: string) => `${serialiseCanvasId(canvasId)}-account-${address}`;
