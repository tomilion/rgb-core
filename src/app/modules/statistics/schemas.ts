export interface TotalTransactionsPayload {
    totalTransactions: number;
}

export const totalTransactionsSchema = {
    $id: "statistics/totalTransactions",
    type: "object",
    required: ["totalTransactions"],
    properties: {
        totalTransactions: { fieldNumber: 1, dataType: "uint32" },
    },
};
