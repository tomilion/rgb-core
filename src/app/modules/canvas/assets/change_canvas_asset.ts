import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from "lisk-sdk";
import { serialiseCanvasId } from "../utils";
import { CanvasPayload, canvasSchema, CanvasState } from "../schemas";

export interface ChangeCanvasPayload {
    canvasId: number;
    costPerPixel?: bigint | null;
    startTime?: bigint | null;
    endTime?: bigint | null;
    width?: number | null;
    height?: number | null;
    timeBetweenDraws?: number | null;
    seed?: bigint | null;
}

export class ChangeCanvasAsset extends BaseAsset<ChangeCanvasPayload> {
    public static readonly ASSET_ID = 1;

    public name = "changeCanvas";
    public id = ChangeCanvasAsset.ASSET_ID;
    public schema = {
        $id: "canvas/changeCanvas-asset",
        title: "ChangeCanvasAsset transaction asset for canvas module",
        type: "object",
        required: ["canvasId"],
        properties: {
            canvasId: { fieldNumber: 1, dataType: "uint32" },
            costPerPixel: { fieldNumber: 2, dataType: "uint64" },
            startTime: { fieldNumber: 3, dataType: "uint64" },
            endTime: { fieldNumber: 4, dataType: "uint64" },
            width: { fieldNumber: 5, dataType: "uint32" },
            height: { fieldNumber: 6, dataType: "uint32" },
            timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
            seed: { fieldNumber: 8, dataType: "uint64" },
        },
    };

    private readonly admin: string;

    public constructor(admin: string) {
        super();
        this.admin = admin;
    }

    public validate({ asset, transaction }: ValidateAssetContext<ChangeCanvasPayload>): void {
        if (transaction.senderAddress.toString("hex") !== this.admin)
        {
            throw new Error("User invalid");
        }

        if ((asset.width !== undefined && asset.width !== null) && (asset.width < 0 || asset.width > 10000))
        {
            throw new Error("Width invalid");
        }

        if ((asset.height !== undefined && asset.height !== null) && (asset.height < 0 || asset.height > 10000))
        {
            throw new Error("Height invalid");
        }

        if ((asset.costPerPixel !== undefined && asset.costPerPixel !== null) && (asset.costPerPixel < 0))
        {
            throw new Error("Cost per pixel invalid");
        }

        if ((asset.seed !== undefined && asset.seed !== null) && (asset.seed < 0))
        {
            throw new Error("Seed invalid");
        }

        const now = BigInt(Math.floor(Date.now() / 1000));

        if ((asset.startTime !== undefined && asset.startTime !== null) && (asset.startTime < now))
        {
            throw new Error("Start time cannot be in the past");
        }

        if ((asset.endTime !== undefined && asset.endTime !== null) && (asset.endTime < now))
        {
            throw new Error("End time cannot be in the past");
        }
    }

    public async apply({ asset, transaction, stateStore }: ApplyAssetContext<ChangeCanvasPayload>): Promise<void> {
        const canvasId = serialiseCanvasId(asset.canvasId);
        const currentCanvas = await stateStore.chain.get(canvasId);

        if (currentCanvas === undefined)
        {
            throw new Error("Canvas does not exist");
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, currentCanvas);

        if (canvas.ownerId.toString("hex") !== transaction.senderAddress.toString("hex"))
        {
            throw new Error("User invalid");
        }

        const now = BigInt(Math.floor(Date.now() / 1000));

        if (canvas.endTime < now || canvas.state === CanvasState.COMPLETE)
        {
            throw new Error("Canvas already ended");
        }

        if (canvas.startTime < now || canvas.state === CanvasState.ACTIVE)
        {
            throw new Error("Canvas already started");
        }

        canvas.width = asset.width ?? canvas.width;
        canvas.height = asset.height ?? canvas.height;
        canvas.costPerPixel = asset.costPerPixel ?? canvas.costPerPixel;
        canvas.startTime = asset.startTime ?? canvas.startTime;
        canvas.endTime = asset.endTime ?? canvas.endTime;
        canvas.timeBetweenDraws = asset.timeBetweenDraws ?? canvas.timeBetweenDraws;
        canvas.seed = asset.seed ?? canvas.seed;

        if (canvas.startTime > canvas.endTime)
        {
            throw new Error("End time must be greater than start time");
        }

        await stateStore.chain.set(canvasId, codec.encode(canvasSchema, canvas));
    }
}
