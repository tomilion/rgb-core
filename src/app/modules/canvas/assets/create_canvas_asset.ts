import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from "lisk-sdk";
import { serialiseCanvasId } from "../utils";
import { canvasSchema, CreateCanvasPayload, PendingPayload, pendingSchema } from "../schemas";

export class CreateCanvasAsset extends BaseAsset<CreateCanvasPayload> {
    public static readonly ASSET_ID = 0;

    public name = "createCanvas";
    public id = CreateCanvasAsset.ASSET_ID;
    public schema = {
        $id: "canvas/createCanvas-asset",
        title: "CreateCanvasAsset transaction asset for canvas module",
        type: "object",
        required: ["canvasId", "costPerPixel", "startBlockHeight", "endBlockHeight", "width", "height", "timeBetweenDraws", "colourPalette", "maxPixelsPerTransaction"],
        properties: {
            canvasId: { fieldNumber: 1, dataType: "uint32" },
            costPerPixel: { fieldNumber: 2, dataType: "uint64" },
            startBlockHeight: { fieldNumber: 3, dataType: "uint64" },
            endBlockHeight: { fieldNumber: 4, dataType: "uint64" },
            width: { fieldNumber: 5, dataType: "uint32" },
            height: { fieldNumber: 6, dataType: "uint32" },
            timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
            colourPalette: { fieldNumber: 8, dataType: "bytes", minLength: 48, maxLength: 48 },
            maxPixelsPerTransaction: { fieldNumber: 9, dataType: "uint32" },
            label: { fieldNumber: 10, dataType: "string", minLength: 0, maxLength: 64 },
        },
    };

    public validate(context: ValidateAssetContext<CreateCanvasPayload>): void {
        const { asset, header } = context;

        if (asset.width < 1 || asset.width > 10000)
        {
            throw new Error("Width invalid");
        }

        if (asset.height < 1 || asset.height > 10000)
        {
            throw new Error("Height invalid");
        }

        if (asset.costPerPixel < 0)
        {
            throw new Error("Cost per pixel invalid");
        }

        if (asset.startBlockHeight < header.height)
        {
            throw new Error("Start block height cannot be in the past");
        }

        if (asset.endBlockHeight < header.height)
        {
            throw new Error("End block height cannot be in the past");
        }

        if (asset.startBlockHeight > asset.endBlockHeight)
        {
            throw new Error("End block height must be greater than start block height");
        }

        if (asset.maxPixelsPerTransaction < 1 || asset.maxPixelsPerTransaction > 10000)
        {
            throw new Error("Max pixels per transaction invalid");
        }

        if (asset.colourPalette.length !== 48)
        {
            throw new Error("Colour palette invalid");
        }
    }

    public async apply(context: ApplyAssetContext<CreateCanvasPayload>): Promise<void> {
        const { asset, stateStore, reducerHandler, transaction } = context;
        const adminAddress = await reducerHandler.invoke<Buffer>("canvas:getAdminAddress");

        if (transaction.senderAddress.compare(adminAddress) !== 0)
        {
            throw new Error("User invalid");
        }

        const canvasId = serialiseCanvasId(asset.canvasId);
        const currentCanvas = await stateStore.chain.get(canvasId);

        if (currentCanvas !== undefined)
        {
            throw new Error("Canvas already exists");
        }

        // Add canvas to state store
        const canvas = {
            ownerId: transaction.senderAddress,
            costPerPixel: asset.costPerPixel,
            startBlockHeight: asset.startBlockHeight,
            endBlockHeight: asset.endBlockHeight,
            width: asset.width,
            height: asset.height,
            timeBetweenDraws: asset.timeBetweenDraws,
            colourPalette: asset.colourPalette,
            maxPixelsPerTransaction: asset.maxPixelsPerTransaction,
            label: asset.label ?? "",
        };
        await stateStore.chain.set(canvasId, codec.encode(canvasSchema, canvas));

        // Add canvas to pending for post block event to determine when to start
        const pendingBuffer = await stateStore.chain.get("canvas:pending");
        const pending = (pendingBuffer !== undefined) ? codec.decode<PendingPayload>(pendingSchema, pendingBuffer) : { canvasIds: [] };
        pending.canvasIds.push(asset.canvasId);
        await stateStore.chain.set("canvas:pending", codec.encode(pendingSchema, pending));
    }
}
