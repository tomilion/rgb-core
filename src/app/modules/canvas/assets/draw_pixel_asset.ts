import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from "lisk-sdk";
import { serialiseAccountId, serialiseCanvasId, serialisePixelId } from "../utils";
import { AccountPayload, accountSchema, CanvasPayload, canvasSchema, CanvasState, pixelSchema } from "../schemas";

export interface DrawPixelPayload {
    canvasId: number;
    coords: number;
    colour: number;
}

export const drawPixelSchema = {
    $id: "canvas/drawPixel/asset",
    title: "DrawPixelAsset transaction asset for Canvas module",
    type: "object",
    required: ["canvasId", "coords", "colour"],
    properties: {
        canvasId: { fieldNumber: 1, dataType: "uint32" },
        coords: { fieldNumber: 2, dataType: "uint32" },
        colour: { fieldNumber: 3, dataType: "uint32" },
    },
};

export class DrawPixelAsset extends BaseAsset<DrawPixelPayload> {
    public static readonly ASSET_ID = 5;

    public name = "drawPixel";
    public id = DrawPixelAsset.ASSET_ID;
    public schema = drawPixelSchema;

    private readonly wallet: string;

    public constructor(wallet: string) {
        super();
        this.wallet = wallet;
    }

    public validate({ asset }: ValidateAssetContext<DrawPixelPayload>): void {
        // Sanity check that coords are within 10000 x 10000 grid
        if (asset.coords < 0 || asset.coords >= 100000000)
        {
            throw Error("Coords invalid");
        }

        // Only first 3 bytes should contain colour information
        if (asset.colour < 0 || asset.colour > 0x00FFFFFF)
        {
            throw Error("Colour invalid");
        }
    }

    public async apply({ asset, transaction, stateStore, reducerHandler }: ApplyAssetContext<DrawPixelPayload>): Promise<void> {
        const canvasId = serialiseCanvasId(asset.canvasId);
        const canvasBuffer = await stateStore.chain.get(canvasId);

        if (canvasBuffer === undefined)
        {
            throw Error("Canvas does not exist");
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);

        if (canvas.state !== CanvasState.ACTIVE)
        {
            throw Error("Canvas not active");
        }

        // Coords encoded as numbered position in canvas eg. (1, 1) in a 1000 x 1000 canvas would be encoded as (1001)
        // as 0 -> 999 is y = 0 and 1000 -> 1999 is y = 1
        if (asset.coords >= (canvas.width * canvas.height))
        {
            throw Error("Coords invalid");
        }

        // Check user has waited for draw timeout before drawing again
        const accountId = serialiseAccountId(asset.canvasId, transaction.senderAddress.toString("hex"));
        const accountBuffer = await stateStore.chain.get(accountId);
        const account = (accountBuffer !== undefined) ? codec.decode<AccountPayload>(accountSchema, accountBuffer) : { lastBlockHeight: 0 };
        const lastBlockHeight = await reducerHandler.invoke<number>("canvas:getLastBlockHeight");
        const currentBlockHeight = lastBlockHeight + 1;

        // TODO: get block time from config
        if ((account.lastBlockHeight + (canvas.timeBetweenDraws / 10)) > currentBlockHeight)
        {
            throw Error("Too many draws");
        }

        await reducerHandler.invoke("token:debit", {
            address: transaction.senderAddress,
            amount: canvas.costPerPixel,
        });

        await reducerHandler.invoke("token:credit", {
            address: Buffer.from(this.wallet, "hex"),
            amount: canvas.costPerPixel,
        });

        const xCoord = asset.coords % canvas.width;
        const yCoord = Math.floor(asset.coords / canvas.height);
        const pixelId = serialisePixelId(asset.canvasId, xCoord, yCoord);

        // Update colour and current owner of pixel
        await stateStore.chain.set(pixelId, codec.encode(pixelSchema, { ownerId: transaction.senderAddress, colour: asset.colour }));

        // Set last draw time
        await stateStore.chain.set(accountId, codec.encode(accountSchema, { lastBlockHeight: currentBlockHeight }));
    }
}
