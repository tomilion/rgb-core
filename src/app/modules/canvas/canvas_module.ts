import { AfterBlockApplyContext, AfterGenesisBlockApplyContext, BaseModule, BeforeBlockApplyContext, codec, TransactionApplyContext } from "lisk-sdk";
import { strict as assert } from "assert";
import { ChangeCanvasAsset } from "./assets/change_canvas_asset";
import { CreateCanvasAsset } from "./assets/create_canvas_asset";
import { DrawPixelAsset, DrawPixelPayload, drawPixelSchema } from "./assets/draw_pixel_asset";
import { serialiseCanvasId, serialisePixelId } from "./utils";
import { ActivePayload, activeSchema, CanvasPayload, canvasSchema, CanvasState, CompletePayload, completeSchema, PendingPayload, pendingSchema, PixelPayload, pixelSchema } from "./schemas";

export class CanvasModule extends BaseModule {
    public static readonly MODULE_ID = 1000;

    public name = "canvas";
    public id = CanvasModule.MODULE_ID;
    public actions = {
        getCanvas: async (params: Record<string, unknown>) => this.getCanvas(params),
        getPendingCanvases: async () => this.getPendingCanvases(),
        getActiveCanvases: async () => this.getActiveCanvases(),
        getCompleteCanvases: async () => this.getCompleteCanvases(),
        getPixels: async (params: Record<string, unknown>) => this.getPixels(params),
    };
    public reducers = {
        getLastBlockHeight: async () => this.getLastBlockHeight(),
    };
    public transactionAssets = [
        new CreateCanvasAsset(process.env.CANVAS_ADMIN_ADDRESS ?? ''),
        new ChangeCanvasAsset(process.env.CANVAS_ADMIN_ADDRESS ?? ''),
        new DrawPixelAsset(process.env.CANVAS_WALLET_ADDRESS ?? ''),
    ];
    public events = ["started", "completed", "pixelChanged"];

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
        // const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(context: AfterBlockApplyContext) {
        const now = BigInt(Math.floor(Date.now() / 1000));

        const pendingToCommit: number[] = [];
        const activeToCommit: number[] = [];

        const pendingBuffer = await context.stateStore.chain.get("canvas:pending");
        const pending = (pendingBuffer !== undefined) ? codec.decode<PendingPayload>(pendingSchema, pendingBuffer) : { canvasIds: [] };

        const activeBuffer = await context.stateStore.chain.get("canvas:active");
        const active = (activeBuffer !== undefined) ? codec.decode<ActivePayload>(activeSchema, activeBuffer) : { canvasIds: [] };

        const completeBuffer = await context.stateStore.chain.get("canvas:complete");
        const complete = (completeBuffer !== undefined) ? codec.decode<CompletePayload>(completeSchema, completeBuffer) : { canvasIds: [] };

        for (const canvasId of pending.canvasIds)
        {
            const canvasBuffer = await context.stateStore.chain.get(serialiseCanvasId(canvasId));
            assert(canvasBuffer !== undefined);
            const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
            assert(canvas.state === CanvasState.PENDING);

            // Automatically starting canvas after passing start time
            if (canvas.startTime <= now)
            {
                canvas.state = CanvasState.ACTIVE;
                await context.stateStore.chain.set(serialiseCanvasId(canvasId), codec.encode(canvasSchema, canvas));
                activeToCommit.push(canvasId);
                this._channel.publish("canvas:started", { canvasId });
                continue;
            }

            pendingToCommit.push(canvasId);
        }

        for (const canvasId of active.canvasIds)
        {
            const canvasBuffer = await context.stateStore.chain.get(serialiseCanvasId(canvasId));
            assert(canvasBuffer !== undefined);
            const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
            assert(canvas.state === CanvasState.ACTIVE);

            // Automatically ending canvas after passing end time
            if (canvas.endTime <= now)
            {
                canvas.state = CanvasState.COMPLETE;
                await context.stateStore.chain.set(serialiseCanvasId(canvasId), codec.encode(canvasSchema, canvas));
                complete.canvasIds.push(canvasId);
                this._channel.publish("canvas:completed", { canvasId });
                continue;
            }

            activeToCommit.push(canvasId);
        }

        await context.stateStore.chain.set("canvas:pending", codec.encode(pendingSchema, { canvasIds: pendingToCommit }));
        await context.stateStore.chain.set("canvas:active", codec.encode(activeSchema, { canvasIds: activeToCommit }));
        await context.stateStore.chain.set("canvas:complete", codec.encode(completeSchema, complete));
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        if (_input.transaction.moduleID === CanvasModule.MODULE_ID &&
            _input.transaction.assetID === DrawPixelAsset.ASSET_ID)
        {
            const pixel = codec.decode<DrawPixelPayload>(drawPixelSchema, _input.transaction.asset);
            this._channel.publish("canvas:pixelChanged", {
                ownerId: _input.transaction.senderAddress.toString("hex"),
                ...pixel,
            });
        }
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccounts = genesisBlock.header.asset.accounts;
    }

    private async getCanvas(params: Record<string, unknown>) {
        const canvasId = serialiseCanvasId(Number(params.canvasId));
        const canvasBuffer = await this._dataAccess.getChainState(canvasId);

        if (canvasBuffer === undefined)
        {
            return null;
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
        return {
            ...canvas,
            ownerId: canvas.ownerId.toString("hex"),
            costPerPixel: Number(canvas.costPerPixel),
            startTime: Number(canvas.startTime),
            endTime: Number(canvas.endTime),
            seed: Number(canvas.seed),
        };
    }

    private async getPendingCanvases() {
        const pendingBuffer = await this._dataAccess.getChainState("canvas:pending");

        if (pendingBuffer === undefined)
        {
            return [];
        }

        return codec.decode<PendingPayload>(pendingSchema, pendingBuffer);
    }

    private async getActiveCanvases() {
        const activeBuffer = await this._dataAccess.getChainState("canvas:active");

        if (activeBuffer === undefined)
        {
            return [];
        }

        return codec.decode<ActivePayload>(activeSchema, activeBuffer);
    }

    private async getCompleteCanvases() {
        const completeBuffer = await this._dataAccess.getChainState("canvas:complete");

        if (completeBuffer === undefined)
        {
            return [];
        }

        return codec.decode<CompletePayload>(completeSchema, completeBuffer);
    }

    private async getPixels(params: Record<string, unknown>) {
        const canvasId = serialiseCanvasId(Number(params.canvasId));
        const canvasBuffer = await this._dataAccess.getChainState(canvasId);

        if (canvasBuffer === undefined)
        {
            return null;
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
        const pixels: number[][] = [];

        for (let y = 0; y < canvas.height; y += 1)
        {
            if (pixels[y] === undefined)
            {
                pixels[y] = [];
            }

            for (let x = 0; x < canvas.width; x += 1)
            {
                const pixelId = serialisePixelId(Number(params.canvasId), x, y);
                const pixelBuffer = await this._dataAccess.getChainState(pixelId);

                if (pixelBuffer === undefined)
                {
                    pixels[y][x] = 0xFFFFFF;
                    continue;
                }

                const pixel = codec.decode<PixelPayload>(pixelSchema, pixelBuffer);
                pixels[y][x] = pixel.colour;
            }
        }

        return pixels;
    };

    private async getLastBlockHeight() {
        const header = await this._dataAccess.getLastBlockHeader();
        return header.height;
    }
}
