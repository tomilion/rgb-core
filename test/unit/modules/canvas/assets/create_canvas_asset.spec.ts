import { codec, testing } from 'lisk-sdk';
import { CreateCanvasAsset, CreateCanvasPayload } from '../../../../../src/app/modules/canvas/assets/create_canvas_asset';
import { CanvasModule } from "../../../../../src/app/modules/canvas/canvas_module";
import { canvasSchema, pendingSchema } from "../../../../../src/app/modules/canvas/schemas";
import { now, numberBetween, randomAddress } from "../../../../utils/random_generator";

describe('CreateCanvasAsset', () => {
    let mockUser: Buffer;
    let mockAsset: CreateCanvasPayload;
    let testClass: CreateCanvasAsset;

    beforeEach(() => {
        mockUser = randomAddress();
        mockAsset = {
            canvasId: numberBetween(0, 4294967295),
            costPerPixel: BigInt(numberBetween(0, 4294967295)),
            startTime: now() + BigInt(numberBetween(0, 604800)),
            endTime: now() + BigInt(numberBetween(1209600, 1814400)),
            width: numberBetween(0, 10000),
            height: numberBetween(0, 10000),
            timeBetweenDraws: numberBetween(0, 4294967295),
            seed: BigInt(numberBetween(0, Number.MAX_SAFE_INTEGER)),
        };
        testClass = new CreateCanvasAsset(mockUser.toString("hex"));
    });

    describe('constructor', () => {
        it('should have valid id', () => {
            expect(testClass.id).toEqual(0);
        });

        it('should have valid name', () => {
            expect(testClass.name).toEqual('createCanvas');
        });

        it('should have valid schema', () => {
            expect(testClass.schema).toMatchSnapshot();
        });
    });

    describe('validate', () => {
        it('should pass with valid args', () => {
            const context = testing.createValidateAssetContext({
                asset: mockAsset,
                transaction: { senderAddress: mockUser } as any,
            });
            testClass.validate(context);
        });

        it('should throw for invalid user', () => {
            const context = testing.createValidateAssetContext({
                asset: mockAsset,
                transaction: { senderAddress: randomAddress() } as any,
            });
            expect(() => testClass.validate(context)).toThrow('User invalid');
        });

        it('should throw width below 0', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, width: -1},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Width invalid');
        });

        it('should throw width above 10000', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, width: 10001},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Width invalid');
        });

        it('should throw height below 0', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, height: -1},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Height invalid');
        });

        it('should throw height above 10000', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, height: 10001},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Height invalid');
        });

        it('should throw cost per pixel below 0', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, costPerPixel: BigInt(-1)},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Cost per pixel invalid');
        });

        it('should throw seed below 0', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, seed: BigInt(-1)},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Seed invalid');
        });

        it('should throw start time in past', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, startTime: now() - BigInt(1), endTime: now() + BigInt(10000)},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('Start time cannot be in the past');
        });

        it('should throw end time in past', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, startTime: now() + BigInt(10000), endTime: now() - BigInt(1)},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('End time cannot be in the past');
        });

        it('should throw start time above end time', () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, startTime: now() + BigInt(10001), endTime: now() + BigInt(10000)},
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow('End time must be greater than start time');
        });
    });

    describe('apply', () => {
        let account: any;
        let chain: any;

        beforeEach(() => {
            account = testing.fixtures.createDefaultAccount([CanvasModule]);
            chain = {};
        });

        it('should save canvas to state store', async () => {
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account]
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: mockAsset,
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });
            jest.spyOn(stateStore.chain, 'set');

            await testClass.apply(context);

            expect(stateStore.chain.set).toHaveBeenCalledWith(
                `canvas-${mockAsset.canvasId}`,
                codec.encode(canvasSchema, {
                    ownerId: account.address,
                    costPerPixel: mockAsset.costPerPixel,
                    startTime: mockAsset.startTime,
                    endTime: mockAsset.endTime,
                    width: mockAsset.width,
                    height: mockAsset.height,
                    timeBetweenDraws: mockAsset.timeBetweenDraws,
                    seed: mockAsset.seed,
                })
            );
            expect(stateStore.chain.set).toHaveBeenCalledWith(
                "canvas:pending",
                codec.encode(pendingSchema, {
                    canvasIds: [mockAsset.canvasId],
                })
            );
        });

        it('should append canvas id to existing pending array', async () => {
            const pending = [numberBetween(0, 4294967295), numberBetween(0, 4294967295), numberBetween(0, 4294967295)];
            chain["canvas:pending"] = codec.encode(pendingSchema, {
                canvasIds: pending,
            });
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: mockAsset,
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });
            jest.spyOn(stateStore.chain, 'set');

            await testClass.apply(context);

            expect(stateStore.chain.set).toHaveBeenCalledWith(
                `canvas-${mockAsset.canvasId}`,
                codec.encode(canvasSchema, {
                    ownerId: account.address,
                    costPerPixel: mockAsset.costPerPixel,
                    startTime: mockAsset.startTime,
                    endTime: mockAsset.endTime,
                    width: mockAsset.width,
                    height: mockAsset.height,
                    timeBetweenDraws: mockAsset.timeBetweenDraws,
                    seed: mockAsset.seed,
                })
            );
            expect(stateStore.chain.set).toHaveBeenCalledWith(
                "canvas:pending",
                codec.encode(pendingSchema, {
                    canvasIds: pending.concat(mockAsset.canvasId),
                })
            );
        });

        it('should throw canvas exists', async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, {
                ownerId: account.address,
                costPerPixel: mockAsset.costPerPixel,
                startTime: mockAsset.startTime,
                endTime: mockAsset.endTime,
                width: mockAsset.width,
                height: mockAsset.height,
                timeBetweenDraws: mockAsset.timeBetweenDraws,
                seed: mockAsset.seed,
            });
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: mockAsset,
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("Canvas already exists");
        });
    });
});
