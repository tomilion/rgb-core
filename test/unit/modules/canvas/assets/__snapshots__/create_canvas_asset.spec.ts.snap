// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCanvasAsset constructor should have valid schema 1`] = `
Object {
  "$id": "canvas/createCanvas-asset",
  "properties": Object {
    "canvasId": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
    "costPerPixel": Object {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
    "endTime": Object {
      "dataType": "uint64",
      "fieldNumber": 4,
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 6,
    },
    "seed": Object {
      "dataType": "uint64",
      "fieldNumber": 8,
    },
    "startTime": Object {
      "dataType": "uint64",
      "fieldNumber": 3,
    },
    "timeBetweenDraws": Object {
      "dataType": "uint32",
      "fieldNumber": 7,
    },
    "width": Object {
      "dataType": "uint32",
      "fieldNumber": 5,
    },
  },
  "required": Array [
    "canvasId",
    "costPerPixel",
    "startTime",
    "endTime",
    "width",
    "height",
    "timeBetweenDraws",
    "seed",
  ],
  "title": "CreateCanvasAsset transaction asset for canvas module",
  "type": "object",
}
`;
